src/caffe/layers/softmax_loss_layer.cpp

```diff
@@ -59,16 +67,19 @@ void SoftmaxWithLossLayer<Dtype>::Forward_cpu(
   int count = 0;
   Dtype loss = 0;
   for (int i = 0; i < outer_num_; ++i) {
-    for (int j = 0; j < inner_num_; j++) {
-      const int label_value = static_cast<int>(label[i * inner_num_ + j]);
-      if (has_ignore_label_ && label_value == ignore_label_) {
-        continue;
+    for (int k = 0; k < label_num_; ++k) {
+      for (int j = 0; j < inner_num_; j++) {
+        const int label_value = static_cast<int>(
+            label[(i * label_num_ + k) * inner_num_ + j]);
+        if (has_ignore_label_ && label_value == ignore_label_) {
+          continue;
+        }
+        DCHECK_GE(label_value, 0);
+        DCHECK_LT(label_value, prob_.shape(softmax_axis_));
+        loss -= log(std::max(prob_data[i * dim + label_value * inner_num_ + j],
```

Comment:
I am not addressing multi-class problem, but multi-label problem where multiple labels are assigned to each instance. In your example, supposing instance <code>i</code> is assigned with 1st and 3rd class, the loss for the instance is simply the average of losses on that two classes, i.e. <code>log(prob_[i * dim + 0 * inner_num_ + j])</code> and <code>log(prob_[i * dim + 2 * inner_num_ + j])</code>. The <code>INNER_PRODUCT</code> still have <code>num_output: 3</code> in this case.
Generally, if there is <code>M</code> classes and each instance has <code>K</code> labels, the shape of <code>data</code> blob is still <code>(N, M)</code> and the shape of <code>label</code> blob is <code>(N, K)</code>. Originally it only allows <code>(N, 1)</code>.